report/src/main/kotlin/com/hartwig/actin/report/pdf/chapters/MolecularDetailsChapter.kt

```diff
@@ -42,86 +43,86 @@ class MolecularDetailsChapter(
 
-        val orangeMolecularTable = Tables.createSingleColWithWidth(contentWidth()).addCell(Cells.createEmpty())
-        report.patientRecord.molecularHistory.latestOrangeMolecularRecord()?.let { molecular ->
-            orangeMolecularTable.addCell(
-                Cells.createTitle("${molecular.experimentType.display()} (${molecular.sampleId}, ${date(molecular.date)})")
-            )
-            if (molecular.hasSufficientQualityButLowPurity()) {
-                val purityString = molecular.characteristics.purity?.let { Formats.percentage(it) } ?: "NA"
-                orangeMolecularTable.addCell(
-                    Cells.createContentNoBorder(
-                        ("Low tumor purity (${purityString}) indicating that potential (subclonal) " +
-                                "DNA aberrations might not have been detected & predicted tumor origin results may be less reliable")
-                    )
-                )
-            }
+        val orangeMolecularRecord = report.patientRecord.molecularHistory.latestOrangeMolecularRecord()
+        val externalPanelResults = report.patientRecord.molecularHistory.molecularTests.filter { it.experimentType == ExperimentType.PANEL }
+        val filteredIhcTests = IHCTestFilter.mostRecentOrUnknownDateIhcTests(report.patientRecord.ihcTests).toList()
+        val groupedByPathologyReport = PathologyReportFunctions.groupTestsByPathologyReport(
+            listOfNotNull(orangeMolecularRecord),
+            externalPanelResults,
+            filteredIhcTests,
+            report.patientRecord.pathologyReports
+        ).filterValues { (orangeTests, molecularTest, ihcTests) ->
+            orangeTests.isNotEmpty() || molecularTest.isNotEmpty() || ihcTests.isNotEmpty()
+        }
 
-            val generators = listOf(MolecularCharacteristicsGenerator(molecular)) + tumorDetailsGenerators(molecular, cohorts, trials)
-            TableGeneratorFunctions.addGenerators(generators, orangeMolecularTable, overrideTitleFormatToSubtitle = true)
+        if (orangeMolecularRecord == null) {
+            document.add(Cells.createContent("No OncoAct WGS and/or Hartwig NGS panel performed"))
+        }
 
-            if (!molecular.hasSufficientQuality) {
-                orangeMolecularTable.addCell(
-                    Cells.createContent(
-                        ("No successful OncoAct WGS and/or tumor NGS panel could be "
-                                + "performed on the submitted biopsy (insufficient quality for reporting)")
-                    )
-                )
+        val table = Tables.createSingleColWithWidth(contentWidth())
+        for ((pathologyReport, tests) in groupedByPathologyReport) {
+            pathologyReport ?: groupedByPathologyReport.keys.takeIf { it.size > 1 }?.let {
+                table.addCell(Cells.createTitle("Other Tests"))
             }
```

Comment:
What is this code supposed to be doing? It seems unusual to only add a cell here when `pathologyReport` is null and the map has >1 key. I think it would be clearer if you performed the `topTable.addCell` from line 106 here as well since the cell is added in both (all?) cases.

Reply:
Thanks for clarifying. What do you think of the following to make the title decision more explicit (I moved the title for known reports here as well):
```suggestion
            when {
                pathologyReport != null -> {
                    table.addCell(Cells.create(PathologyReportFunctions.getPathologyReportSummary(report = pathologyReport)))
                }
                groupedByPathologyReport.size > 1 -> {
                    table.addCell(Cells.createTitle("Other Tests"))
                }
                else -> {} // No title needed: all tests are ungrouped
            }
```
You could even drop the `else` case; I just get a little wary about conditional cell additions since it can lead to malformed tables in some cases.